### PR TITLE
SNOW-683059 Review test coverage for column.py, dataframe.py, etc

### PR DESCRIPTION
**Pull request by sfc-gh-stan**
Please answer these questions before submitting your pull requests. Thanks!

1. What GitHub issue is this PR addressing? Make sure that there is an accompanying issue to your PR.

   Fixes #683059

2. Fill out the following pre-review checklist:

   - [ ] I am adding a new automated test(s) to verify correctness of my new code
   - [ ] I am adding new logging messages
   - [ ] I am adding a new telemetry message
   - [ ] I am adding new credentials
   - [ ] I am adding a new dependency

3. Please describe how your code solves the related issue.

### CODE REVIEW

File: src/snowflake/snowpark/column.py [hunk 1]
```diff
@@ -605,7 +605,7 @@ def __str__(self):
         return f"Column[{self._expression}]"
 
     def __repr__(self):
-        return f"Column({self._expression})"
+        return f"Column({self._expression})"  # pragma: no cover
```
Comment (sfc-gh-stan) [inline src/snowflake/snowpark/column.py:608]: The onlt test I could think of for `__repr__` is `assert eval(repr(x)) == x`, which does not apply for our case.

File: src/snowflake/snowpark/file_operation.py [hunk 2]
```diff
@@ -219,8 +219,7 @@ def get(
                     self._session, plan
                 )._internal_collect_with_tag(statement_params=statement_params)
             return [GetResult(**file_result.asDict()) for file_result in get_result]
-        # connector raises IndexError when no file is downloaded from python connector.
-        except IndexError:
+        except OperationalError:
```
Comment (sfc-gh-stan) [inline src/snowflake/snowpark/file_operation.py:222]: The second test case in `test_get_negative_test` fails with OperationalError instead of IndexError. I looked into Python connector's code and couldn't find where it raises `IndexError` in file transfer related code.

Comment (sfc-gh-stan) [reply src/snowflake/snowpark/file_operation.py:222]: Also:  Looks like `test_get_negative_test` is skipped to avoid Azure/GCP-specific issues, we should still run it for AWS. Let's follow Python connector and use markers `@pytest.mark.aws` to select the corresponding tests in our tox config.

File: src/snowflake/snowpark/column.py [hunk 3]
```diff
@@ -81,7 +81,7 @@ def _to_col_if_lit(
         return col
     elif isinstance(col, VALID_PYTHON_TYPES_FOR_LITERAL_VALUE):
         return Column(Literal(col))
-    else:
+    else:  # pragma: no cover
```
Comment (sfc-gh-sfan) [inline src/snowflake/snowpark/column.py:84]: Frankly I think these `else` cases should be fairly simple to test in a unit test, though given they are "private" functions, it is probably okay to skip.

File: src/snowflake/snowpark/column.py (hunk 1 above)
Comment (sfc-gh-sfan) [inline src/snowflake/snowpark/column.py:608]: Why is `__repr__` different from `__str__`? Also this might be testable?

File: src/snowflake/snowpark/dataframe.py [hunk 4]
```diff
@@ -813,7 +809,7 @@ def to_df(self, *names: Union[str, Iterable[str]]) -> "DataFrame":
         """
         col_names = parse_positional_args_to_list(*names)
         if not all(isinstance(n, str) for n in col_names):
-            raise TypeError(
+            raise TypeError(  # pragma: no cover
```
Comment (sfc-gh-sfan) [inline src/snowflake/snowpark/dataframe.py:812]: This is an API call. Should we have a negative test?

File: src/snowflake/snowpark/dataframe.py [hunk 5]
```diff
@@ -840,12 +836,12 @@ def __getitem__(self, item):
         elif isinstance(item, int):
             return self.__getitem__(self.columns[item])
         else:
-            raise TypeError(f"Unexpected item type: {type(item)}")
+            raise TypeError(f"Unexpected item type: {type(item)}")  # pragma: no cover
```
Comment (sfc-gh-sfan) [inline src/snowflake/snowpark/dataframe.py:839]: I feel this should be tested?

File: src/snowflake/snowpark/dataframe.py [hunk 6]
```diff
@@ -840,12 +836,12 @@ def __getitem__(self, item):
...
 
     def __getattr__(self, name):
         # Snowflake DB ignores cases when there is no quotes.
         if name.lower() not in [c.lower() for c in self.columns]:
-            raise AttributeError(
+            raise AttributeError(  # pragma: no cover
```
Comment (sfc-gh-sfan) [inline src/snowflake/snowpark/dataframe.py:844]: Is this testable?

File: src/snowflake/snowpark/dataframe.py [hunk 7]
```diff
@@ -956,7 +952,7 @@ def select(
                 )
                 names.extend(new_cols)
             else:
-                raise TypeError(
+                raise TypeError(  # pragma: no cover
```
Comment (sfc-gh-sfan) [inline src/snowflake/snowpark/dataframe.py:955]: This is an API call. Should we have a negative test?

File: src/snowflake/snowpark/dataframe.py [hunk 8]
```diff
@@ -1038,7 +1034,7 @@ def drop(
         """
         # an empty list should be accept, as dropping nothing
         if not cols:
-            raise ValueError("The input of drop() cannot be empty")
+            raise ValueError("The input of drop() cannot be empty")  # pragma: no cover
```
Comment (sfc-gh-sfan) [inline src/snowflake/snowpark/dataframe.py:1037]: This is an API call. Should we have a negative test?

File: src/snowflake/snowpark/dataframe.py [hunk 9]
```diff
@@ -2006,7 +2002,7 @@ def join(
             elif not isinstance(using_columns, list):
                 raise TypeError(
                     f"Invalid input type for join column: {type(using_columns)}"
-                )
+                )  # pragma: no cover
```
Comment (sfc-gh-sfan) [inline src/snowflake/snowpark/dataframe.py:2005]: This is an API call. Should we have a negative test?

File: src/snowflake/snowpark/dataframe.py [hunk 10]
```diff
@@ -2016,7 +2012,7 @@ def join(
                 rsuffix=rsuffix,
             )
 
-        raise TypeError("Invalid type for join. Must be Dataframe")
+        raise TypeError("Invalid type for join. Must be Dataframe")  # pragma: no cover
```
Comment (sfc-gh-sfan) [inline src/snowflake/snowpark/dataframe.py:2015]: This is an API call. Should we have a negative test?

File: src/snowflake/snowpark/dataframe.py [hunk 11]
```diff
@@ -2869,7 +2865,7 @@ def create_or_replace_view(
         elif isinstance(name, (list, tuple)) and all(isinstance(n, str) for n in name):
             formatted_name = ".".join(name)
         else:
-            raise TypeError(
+            raise TypeError(  # pragma: no cover
```
Comment (sfc-gh-sfan) [inline src/snowflake/snowpark/dataframe.py:2868]: This is an API call. Should we have a negative test?

File: src/snowflake/snowpark/dataframe.py [hunk 12]
```diff
@@ -3176,7 +3172,9 @@ def with_column_renamed(self, existing: ColumnOrName, new: str) -> "DataFrame":
                     f"Unable to rename column {existing} because it doesn't exist."
                 )
         else:
-            raise TypeError("'exisitng' must be a column name or Column object.")
+            raise TypeError(
```
Comment (sfc-gh-sfan) [inline src/snowflake/snowpark/dataframe.py:3175]: This is an API call. Should we have a negative test?

File: src/snowflake/snowpark/dataframe.py [hunk 13]
```diff
@@ -3374,7 +3372,7 @@ def _explain_string(self) -> str:
             if exec_plan:
                 msg = f"{msg}\nLogical Execution Plan:\n{exec_plan}"
             else:
-                msg = f"{self._plan.queries[0].sql} can't be explained"
+                msg = f"{self._plan.queries[0].sql} can't be explained"  # pragma: no cover
```
Comment (sfc-gh-sfan) [inline src/snowflake/snowpark/dataframe.py:3375]: I wonder when this would happen

File: src/snowflake/snowpark/column.py (hunk 1 above)
Comment (sfc-gh-stan) [reply src/snowflake/snowpark/column.py:608]: [`__repr__` vs `__str__`](https://stackoverflow.com/questions/1436703/what-is-the-difference-between-str-and-repr)
`assert eval(repr(x)) == x` does not apply for this case, so I chose to skip.

Comment (sfc-gh-sfan) [reply src/snowflake/snowpark/column.py:608]: Makes sense. I thought using parenthesis is not too bad, but anyways this is okay :)

File: src/snowflake/snowpark/dataframe.py (hunk 12 above)
Comment (sfc-gh-stan) [reply src/snowflake/snowpark/dataframe.py:3175]: I don't think we need to test for negative cases of input that are indicated by the type hints already, but I'm open to discussion :)

File: src/snowflake/snowpark/dataframe.py [hunk 14]
```diff
@@ -1038,7 +1034,7 @@ def drop(
         """
         # an empty list should be accept, as dropping nothing
```
Comment (sfc-gh-stan) [inline src/snowflake/snowpark/dataframe.py:1035]: This comment is contradicting the sanity check on line 1037 below 😕

File: src/snowflake/snowpark/column.py (hunk 1 above)
Comment (sfc-gh-jdu) [reply src/snowflake/snowpark/column.py:608]: Should we change to use `__str__`. It shouldn't be a breaking change

File: src/snowflake/snowpark/dataframe_na_functions.py [hunk 15]
```diff
@@ -127,6 +127,14 @@ def drop(
             |4.0  |NULL  |
             --------------
             <BLANKLINE>
+            >>> df.na.drop(subset="a").show()
```
[the remark this answers is not in the document]
Comment (sfc-gh-jdu) [reply src/snowflake/snowpark/dataframe_na_functions.py:130]: Dropping support for this case will be a behavior change. Let's just fix the type hints. We should also support it for spark migration.

File: src/snowflake/snowpark/file_operation.py (hunk 2 above)
Comment (sfc-gh-jdu) [reply src/snowflake/snowpark/file_operation.py:222]: This will be a behavior change

File: src/snowflake/snowpark/dataframe_na_functions.py [hunk 16]
```diff
@@ -137,7 +145,9 @@ def drop(
         # iff(non_float_col is null, 0, 1) >= thresh
 
         if how is not None and how not in ["any", "all"]:
-            raise ValueError("how ('" + how + "') should be 'any' or 'all'")
+            raise ValueError(
+                "how ('" + how + "') should be 'any' or 'all'"
```
Comment (sfc-gh-jdu) [inline src/snowflake/snowpark/dataframe_na_functions.py:149]: use f-string here?

File: src/snowflake/snowpark/dataframe.py (hunk 6 above)
Comment (sfc-gh-jdu) [reply src/snowflake/snowpark/dataframe.py:844]: Yea we should have negative test for this one

File: src/snowflake/snowpark/dataframe.py (hunk 12 above)
Comment (sfc-gh-sfan) [reply src/snowflake/snowpark/dataframe.py:3175]: Oh I'm not aware of that lol. Github just shows this line to me. It does not hurt to test but yeah it's also fine to leave it as is. No strong opinion then.

File: src/snowflake/snowpark/dataframe_na_functions.py [hunk 17]
```diff
@@ -146,7 +146,7 @@ def drop(
 
         if how is not None and how not in ["any", "all"]:
             raise ValueError(
```
Comment (sfc-gh-aalam) [inline src/snowflake/snowpark/dataframe_na_functions.py:148]: I think we should add a negative test for checking this as well

File: src/snowflake/snowpark/file_operation.py (hunk 2 above)
Comment (sfc-gh-stan) [reply src/snowflake/snowpark/file_operation.py:222]: `except OperationalError`? But I think this is the intended behavior: when no files are downloaded, just return a empty list and don't raise exceptions.

Comment (sfc-gh-jdu) [reply src/snowflake/snowpark/file_operation.py:222]: yea, changing the exception type is usually a behavior change because some users might use try/except to catch the expected exception type.

Comment (sfc-gh-yixie) [reply src/snowflake/snowpark/file_operation.py:222]: I suggest keeping the IndexError there. I remember I've seen IndexError before.

File: src/snowflake/snowpark/dataframe.py (hunk 13 above)
Comment (sfc-gh-yixie) [reply src/snowflake/snowpark/dataframe.py:3373]: @sfc-gh-jdu Do you know what sql can't be explained?

File: src/snowflake/snowpark/column.py (hunk 1 above)
Comment (sfc-gh-yixie) [reply src/snowflake/snowpark/column.py:608]: Makes sense. The problem is `Expression` doesn't override `__repr__`.

File: src/snowflake/snowpark/file_operation.py (hunk 2 above)
Comment (sfc-gh-stan) [reply src/snowflake/snowpark/file_operation.py:222]: Okay, I will revert this change for now, [SNOW-687929](https://snowflakecomputing.atlassian.net/browse/SNOW-687929) is created to track this fix. [SNOW-687933](https://snowflakecomputing.atlassian.net/browse/SNOW-687933) is created to track the proposal of using cloud provider specific pytest marks.

**Review by sfc-gh-jdu — APPROVED** [review]
